apps/demo-app/pages/api/check-signature.ts

```diff
@@ -0,0 +1,205 @@
+ *
+ * @param address - The address that is supposed to have signed the message.
+ * @param pubKey - The public key associated with the session address base64 encoded.
+ * @param messageString - The message that is supposed to have been signed.
+ * @param signature - The signature to verify against the message and address.
+ * @returns True if the signature is valid, false otherwise.
+ */
+export function verifyXionSignature(
+  address: string,
+  pubKey: string,
+  messageString: string,
+  signature: string,
+): boolean {
+  const signatureBuffer = Buffer.from(signature, "base64");
+  const uint8Signature = new Uint8Array(signatureBuffer); // Convert the buffer to an Uint8Array
+  const pubKeyValueBuffer = Buffer.from(pubKey, "base64"); // Decode the base64 encoded value
+  const pubKeyUint8Array = new Uint8Array(pubKeyValueBuffer); // Convert the buffer to an Uint8Array
+
+  return verifyADR36Amino(
+    "xion",
+    address,
+    messageString,
+    pubKeyUint8Array,
+    uint8Signature,
+  );
+}
+
+/**
+ * Verifies the Xion signature and grants for a given address, session address, public key,
+ * message string, and signature.
+ *
+ * @param address - The address to verify the grants for.
+ * @param sessionAddress - The session address to verify the grants for.
+ * @param pubKey - The public key associated with the session address base64 encoded.
+ * @param messageString - The message string to verify the signature against.
+ * @param signature - The signature to verify.
+ *
+ * @returns Promise<boolean> - A promise that resolves to true if the Xion signature and grants are valid,
+ *                             or false otherwise.
+ */
+export async function verifyXionSignatureAndGrants(
+  address: string,
+  sessionAddress: string,
+  pubKey: string,
+  messageString: string,
+  signature: string,
+): Promise<boolean> {
+  const isValid = verifyXionSignature(
+    sessionAddress,
+    pubKey,
+    messageString,
+    signature,
+  );
+  if (!isValid) {
+    return false;
+  }
+
+  return verifyGrants(address, sessionAddress);
+}
+
+/**
+ * Verifies if grants have been given by a granter to a grantee.
+ *
+ * @param grantee - The address of the granter.
+ * @param granter - The address of the grantee.
+ *
+ * @returns Promise<boolean> - A promise that resolves to a boolean indicating whether ANY grants have been given.
+ */
+export async function verifyGrants(
+  granter: string,
+  grantee: string,
+): Promise<boolean> {
+  const baseUrl = `${testnetChainInfo.rest}/cosmos/authz/v1beta1/grants`;
+  const url = new URL(baseUrl);
+  const params = new URLSearchParams({
+    grantee,
+    granter,
+  });
+  url.search = params.toString();
+  const data = await fetch(url, {
+    cache: "no-store",
+  })
+    .then((response): Promise<QueryGrantsResponse> => response.json())
+    .catch((err) => {
+      console.error("Could not fetch grants info", err);
+    });
```

Comment:
Seems like we're mixing async/await syntax and promise .then chaining. Not an immediate issue but was tough to parse at first.